### PR TITLE
Improve performance and reliability

### PR DESCRIPTION
**Pull request by yakkomajuri**
Dear RudderStack friends, Yakko from the PostHog Extensibility Team here!

This plugin is great, I took it for a spin and worked exactly as intended.

It was built before we added some new functionality to our plugin server though, which can help improve its performance. 

There was no way you could've built it using these features because they didn't exist yet! However, if you don't mind, I've taken the time to implement them. They will make the plugin more performant and reliable.

Here's a summary of what this PR does (I've also added comments to the code to help):

### Buffer

Using `createBuffer` I've made it so that payload uploads happen in the background, so that they don't stall our ingestion pipeline.

### onEvent

We've introduced a function called `onEvent` exactly for plugins like these that export events but don't modify them. Here, I've made it so that when `onEvent` gets called, we add the event to the buffer to be uploaded later (maximum 5s later).

### jobs

I've also used jobs to implement a retry queue that is more robust. Using the cache as is currently done can cause some concurrency issues. The new retry queue uses jobs in order to try to upload payloads that failed up to 15 times within a 24h period, so that data loss is very unlikely.

As a result, `fetchWithRetry` wasn't needed anymore either.

### console.log

I've also reduced the amount of console logs because this plugin was a bit too noisy for production, and I'd love to make it available to all of our users!

---

If there are any questions, feel free to let me know!

### CODE REVIEW

**Comment by yakkomajuri** [comment]
Accidentally ran prettier over the entire repo, sorry about that. Happy to change it

**Comment by yakkomajuri** [comment]
cc @sayan-mitra @ameypv-rudder

File: index.js
Comment (neilkakkar) [inline index.js:275]: So we stick to the 100kB limit, add the size of the payload (default is 1 byte otherwise, which isn't ok)

```suggestion
    global.buffer.add(rudderPayload, JSON.stringify(rudderPayload).length)
```

**Comment by arnab-p** [comment]
@yakkomajuri Thanks for taking the time to improve this. We'll quickly test the PR at our end and get back to you.

**Comment by utsabc** [comment]
Hey, @yakkomajuri thanks for taking the time to improve the functionality of the plugin. I have checked the new implementation based on the new features which are added for plugins as documented [Here](https://posthog.com/docs/plugins/build#pluginmeta). I am trying to use the plugin to ingest data from my local Posthog instance (Docker Based) to Rudder but data are not getting ingested to Rudder, maybe I am missing something here. 

I tested out with the existing implementation of the plugin which is working though. Is it possible for you to jump into a call sometime to guide us through?

Thanks in advance

cc @arnab-p

**Comment by yakkomajuri** [comment]
Hey @utsabc I'll double check and get back to you. I managed to fill another PostHog instance via Rudderstack using this plugin.

If everything works for me, we can hop on a call! Also, what version of PostHog are you on?

**Comment by utsabc** [comment]
> Hey @utsabc I'll double check and get back to you. I managed to fill another PostHog instance via Rudderstack using this plugin.
> 
> If everything works for me, we can hop on a call! Also, what version of PostHog are you on?

I am using the docker-compose.yml file present in the posthog repository to run the instance. Currently, it's pulling the posthog docker image with the latest tag (`posthog/posthog:latest`)

**Comment by utsabc** [comment]
Hey, @yakkomajuri  any updates on this?

**Comment by yakkomajuri** [comment]
Nope, sorry, been super busy. Hopefully getting to this today

**Comment by arnab-p** [comment]
@yakkomajuri, I Wanted to follow up on the thread to check the progress here. But, unfortunately, we're currently blocked on this.

**Comment by yakkomajuri** [comment]
On it at the moment.

**Comment by yakkomajuri** [comment]
Just tested this and it worked perfectly.

I'm running PostHog locally and sending the events to RudderStack Cloud and then into a PostHog Cloud instance. 

It does take a little time for the events to show up though, as it batches and processes them asynchronously.

**Comment by yakkomajuri** [comment]
Here's a video of me using the plugin. It seems there are some timeout errors, but that appears to be something on the RudderStack end (probably because of the limits on my free trial). The events that fail to send nevertheless get retried and succeed later as expected. I did however change the limits on the buffer to see if that helps with the errors.


https://user-images.githubusercontent.com/38760734/122936278-696fb780-d347-11eb-9293-3ee92a6a0139.mov

**Comment by utsabc** [comment]
> Here's a video of me using the plugin. It seems there are some timeout errors, but that appears to be something on the RudderStack end (probably because of the limits on my free trial). The events that fail to send nevertheless get retried and succeed later as expected. I did however change the limits on the buffer to see if that helps with the errors.
> rudder.plugin.mov

Thanks, @yakkomajuri  for checking this out thoroughly, and explaining it. Maybe the nature of my pipe was causing the issue, the setup goes like this.

Batch of events (Http Source) --> Rudder --> Posthog (Local Instance) --> Rudder Again (Via Plugin) --> Webhook

Let me recheck with this setup once, though I am not sure if the way I am setting up the Pipe should cause issues.

**Comment by utsabc** [comment]
@yakkomajuri  @neilkakkar  We are all good with this PR. @neilkakkar  Can you please Approve this from your end and add me, and @arnab-p  as approver.

PS. I mistakenly commited 2 commits to this PR while I was testing it locally extremely sorry for that. I have reverted those so no worries

**Review by neilkakkar — APPROVED** [review]

**Comment by neilkakkar** [comment]
Hey @utsabc - I can approve it, but can't add you as an approver, since no access on rudderlabs repo! But, you should be able to do that?

File: index.js
Comment (neilkakkar) [inline index.js:235]:
```suggestion
    // Setup a buffer to group events to be sent to RudderStack in the background at most every 60s
```

**Review by utsabc — APPROVED** [review]
All good

**Comment by arnab-p** [comment]
@yakkomajuri Thanks for your contribution. Can you please sign the CLA [here](https://rudderlabs.wufoo.com/forms/rudderlabs-contributor-license-agreement)?

**Comment by yakkomajuri** [comment]
Hey @arnab-p! This is an MIT-licensed repo, so if you wanted to go ahead and commit the changes from my fork you'd be able to. I'm not comfortable signing a CLA though.

**Comment by arnab-p** [comment]
As suggested by @yakkomajuri, we've contributed to this [PR](https://github.com/rudderlabs/rudderstack-posthog-plugin/pull/3) and closing this PR.

**Pull request closed without merging**
